### PR TITLE
Start drafting an updated pep.

### PR DESCRIPTION
**Pull request by mikeshardmind**
This is not ready, so much more detail and examples still need to be added. The parts that are here already and not dependent on a particular TODO note are ready for any constructive criticism.

Since last major set of discussion, the only change in overall direction was taking feedback expressed on discord that led to keeping the ordered construct, but advocating for the more ergonomic option with `Intersection` and `&`

### CODE REVIEW

**Comment by mark-todd** [comment]
@mikeshardmind Exciting to see the work-in-progress! Just an initial thought re. the new structure of A & B. If I've understood it correctly, I quite like this solution. `Intersection` must be used for any cases that would create issue or ambiguity, but `&` may be used if the intersection is unambiguous in an unordered form.

I did think of one of more case that I think would probably need the same treatment as `Any` and `TypeVarTuple`, which is `TypeVar`. Consider:

```py
T = TypeVar("T")
def func(x: T) -> T & X:
    pass
y: Any = 1
func(y)
```
My point is a TypeVar can be substituted with anything and create one of the ambiguous types in some subcases.

**Comment by mikeshardmind** [comment]
@mark-todd 
> Intersection must be used for any cases that would create issue or ambiguity, but & may be used if the intersection is unambiguous in an unordered form.

This requirement was dropped after feedback from Jelle in Discord prior to turning the draft into a PR draft, the rules for ambiguity are still included as a set of things type checkers may want to offer to users as a configurable rule that is not strictly about correctness (for comparison, see many of mypy and pyright's rules that prevent *accidental* uses of Any)

> I did think of one of more case that I think would probably need the same treatment as Any and TypeVarTuple, which is TypeVar.

This one doesn't need it. When the user is passing in `Any` and getting back `Any & X`, the user started out with a type that allows all the same things as the one they got back, there isn't new ambiguity for the caller, and there wasn't ambiguity for the function body either, as the function always has the responsibility there of creating a valid type that could be mixed with anything (or erroring if it can't)

It's covered in what is (at least currently) the only appendix, but isn't a matter for correctness for the specification.

**Comment by mark-todd** [comment]
@mikeshardmind Ok sure I see your point about `Any & X`, but what about non-LSP violating method overwrites? Apologies if I've misremembered any of the previous discussion, but E.g.:

```py
class X:
    def foo(x: int) -> int:
        ...

class Y:
    def foo(y: str) -> str:
        ...

T = TypeVar("T")

def func(a: T) -> T & X:
    ...

x = func(Y())
```

In our ordered version, we simply say Y takes precendence over X, but in the unordered one I'm not sure how we resolve it.

**Comment by mark-todd** [comment]
> This requirement was dropped after feedback from Jelle in Discord prior to turning the draft into a PR draft, the rules for ambiguity are still included as a set of things type checkers may want to offer to users as a configurable rule that is not strictly about correctness (for comparison, see many of mypy and pyright's rules that prevent accidental uses of Any)

@mikeshardmind Ah ok, perhaps my comment isn't relevant now - I'll leave the above in case the train of thought proves useful.

**Comment by mikeshardmind** [comment]
> but in the unordered one I'm not sure how we resolve it.

It's still an ordered construct, we're just keeping it shorter in name.

**Comment by mark-todd** [comment]
> It's still an ordered construct, we're just keeping it shorter in name.

Oh ok, if it's exactly equivalent should be fine - I suppose the appendix you mentioned should perhaps provide a list of subcases to be aware of, but maybe we can leave this to the makers of type checkers.

**Comment by mikeshardmind** [comment]
All further progress on hold pending discussion here: https://discuss.python.org/t/rules-for-subclassing-of-any/47604 Impacts on consistency of this significantly change the calculus for intersections.

**Comment by DiscordLiz** [comment]
Am I right in seeing that your option 5 from before would be the most consistent with what already exists if what type checkers are doing now is codified?

**Comment by mikeshardmind** [comment]
@DiscordLiz 
I may need to reevaluate based on the outcome of that discussion, remaining as consistent with both the specification and our prior discussions as possible, but this could lead to needing to go back to options which were previously excluded for not being consistent with the existing world.

The assumptions the discussions had about Any & T assumed that Any *can* violate LSP, and that Any's position in the MRO mattered. Both of those are now up for larger input from the typing community.

**Comment by mikeshardmind** [comment]
Initial evaluation on that to keep busy, there's 4 options in total is the answer is a simple "yes/no" for both parts.

If what type checkers are doing there is codified, we need to go back to option 5 to remain consistent with the rest of the type system.

If LSP is retained, but MRO becomes respected again, I can continue working with the current approach and be consistent with the rest of the type system.

if MRO is respected, but LSP is thrown out the window by Any, I'll need more time to think about useful definitions. There's at lease some consistent ones.

if MRO remains ignored for Any, and LSP is thrown out on top of it, I don't think we have a consistent definition available.

File: docs/historical_specification.rst [hunk 1]
```diff
@@ -0,0 +1,668 @@
...
+Every Callable within an intersection shall be treated like a ``def __call__()`` Protocol.
+
+::
+
+    from typing import Protocol, Callable
+
+    MyCallable = Callable[[str, int], float]
+
+    class CallProto:
+        def __call__(a: str, b: int) -> float: ...
```
Comment (stroxler) [inline docs/historical_specification.rst:522]:
```suggestion
        def __call__(a: str, b: int, /) -> float: ...
```

Might be good to make the callable protocol positional-only; with normal arguments `a` and `b` the types aren't exactly equivalent.

**Review by stroxler — COMMENTED** [review]
This was a great read, thanks for putting it up!

File: docs/specification.rst [hunk 2]
```diff
@@ -1,668 +1,481 @@
...
+
+
+Specification
+=============
+
+This adds a type form to the ``typing`` module named ``Intersection``
+and implements ``__and__`` for ``builtins.type``. ``Intersection`` is a type
+form (todo defintion) that can be parametrized with with type expressions.
+
+This exands the allowed use of ``typing.Self`` (hereafter ``Self``) for type
```
Comment (mark-todd) [inline docs/specification.rst:136]: *expands

File: docs/historical_specification.rst [hunk 3]
```diff
@@ -0,0 +1,668 @@
...
+As `PEP 483`_ already suggested: ``Any`` shall be removed from an ``Intersection``, i.e.
+``Intersection[A, B, Any] == Intersection[A, B]``.
+
+% This is only a suggestion and needs to be discussed and decided in https://github.com/CarliJoy/intersection_examples/issues/1
+% Once it was finally decided the discussion and arguments should be summarized here.
+
+
+``Never`` Evaluation
+--------------------
+An intersection that contains either two classes that are a or are a subclass of two different `internal base classes <https://docs.python.org/3/library/stdtypes.html>`_ shall evaluate to ``Never``.
```
[the remark this answers is not in the document]
Comment (mikeshardmind) [reply docs/historical_specification.rst]: This was moved from specification -> historical_specification as a means to help people view what's changed from when we started to now side by side in tools with rst rendering support since there has been too much discussion for people to catch up on in a reasonable amount of reading time. I have no intent on trying to determine prior intent for what was there before and maintain it. The currently relevant draft is still at specification.rst, apologies for any confusion that means of going about it may have caused.

I've just gone ahead and removed the file entirely, people using github's web interface can make that comparison by opening up different points in history, people comparing locally can use git worktree.

**Comment by mikeshardmind** [comment]
Update for those not closely following the other related discussion on discourse:

There appear to be 2 or 3 viable paths forward for Any as a subclass in basic inheritance (2 behaviorally, but 3 ways of reaching one of the two)

1 of those 3 would not require changes to #38 (the method the draft is currently written under)
1 of those 3 would have one way forward, the prior option 5 from the much longer #31 
1 of those 3 would be mostly the same as current, but sort Any at the end no matter where it exists in the order *so as long as that is the behavior of basic subclassing* (allowing removing only that part if the same stops happening in normal subclassing MRO resolution)

I'll continue working on the draft as I have time assuming that either the first or last of those happens as they seem the most likely outcomes, and have the most in common with each other, but I don't think it should be submitted as a PEP draft prior to the more basic behavior being chosen.

File: docs/specification.rst [hunk 4]
```diff
@@ -1,668 +1,481 @@
...
+When comparing what other languages with intersections, most do not provide an
+intersection type.
+
+Two notable comparisons with languages that do reinforce the decision to
+include ordering.
+
+While Kotlin (unlike Java) allows multiple inheritence, it does not allow a
+diamond pattern to exist, allowing only one base to implement a defintion for
+a method or attribute of a type. Kotlin's intersections therefore do not have
+to consider non-disjoint intersections.
```
Comment (DiscordLiz) [inline docs/specification.rst:372]: When not targeting the JVM, Kotlin's `dynamic` is available. Synthesizing bounds for dynamic is not something currently done: https://github.com/JetBrains/kotlin/blob/master/spec-docs/dynamic-types.md you may want to specify Kotlin, while targeting the JVM or reword this to account for dynamic better.

Comment (mikeshardmind) [reply docs/specification.rst:372]: I'm not sure we need to bring this part into the discussion. I'll think about it more, but the prior part sticks to why it's the case, rather than an example of how.

it might also be worth adding that while Kotlin has intersection types, it has them through smart casting and they are not directly user-denotable 

They arise with

```kt
fun process(item: Any) {
    if (item is A && item is B) {
        // smart cast to an intersection of A and B
    }
}
```

or (compile time checked that use of the function is as expected)

```kt
fun <T> process(item: T) where T: A, T: B {
    // item is an intersection of A and B
}
```

But expressed as bounds on a type.

**Comment by mark-todd** [comment]
> All further progress on hold pending discussion here: https://discuss.python.org/t/rules-for-subclassing-of-any/47604 Impacts on consistency of this significantly change the calculus for intersections.

> I'll continue working on the draft as I have time assuming that either the first or last of those happens as they seem the most likely outcomes, and have the most in common with each other, but I don't think it should be submitted as a PEP draft prior to the more basic behaviour being chosen.

@mikeshardmind I've just noticed the thread on this is now closed from timeout, so it certainly feels to me at least that the discussion has come to an end. Do we have a conclusion on how to proceed? If not I propose we assume option 1 you listed here is the behaviour, and recircle to it at a later date if needed. Ultimately if this does prove to be a point that needs resolving I'm sure it'll reappear in a later PEP discussion anyway. Might be worth adding an example to the PEP that explains the assumed behaviour to avoid doubt.

**Comment by mark-todd** [comment]
Btw I know this is a small point but - could we add a short README or link describing what package to render the .rst? I've tried a few different packages but none of them look quite right - is there a specific one for PEPs?

File: docs/specification.rst [hunk 5]
```diff
@@ -1,668 +1,481 @@
...
+        	    return self.handle_no_permission()  # Valid
+            # calling a method of `View`
+            return super().dispatch(request, *args, **kwargs)  # Valid
+
+    invalid = LoginRequiredMixin()  # error here
+    # invalid now has an instance method where self is bound to a value
+    # inconsistent with it's annotation.
+
+Source: https://docs.djangoproject.com/en/3.2/topics/auth/default/#django.contrib.auth.mixins.LoginRequiredMixin
+
```
Comment (mark-todd) [inline docs/specification.rst:127]: The `Self & View` example I find a little confusing here, although I understand the need for replicating real world cases. Wondering if maybe a mixin case based on something like this might be better as a first one? Just to warm people up to 
 the concepts:
https://stackoverflow.com/questions/70698000/workaround-for-lack-of-intersection-types-with-python-generics

**Comment by mikeshardmind** [comment]
@mark-todd discussion restarted at https://discuss.python.org/t/take-2-rules-for-subclassing-any/47981 per Eric's comment on the immediate back and forth being harder to follow

File: docs/specification.rst [hunk 6]
```diff
@@ -1,668 +1,481 @@
...
+Given an ``Intersection[*Ts]`` as a type expression, a corresponding value
+expression is considered to be consistent with it if and only if the value
+expression would be consistent with all of the type expressions, ``*Ts``.
+
+Given a value that has a type of ``Intersection[*Ts]``, use of the value is
+consistent with the known type if at least one type in ``*Ts`` provides a
+definition which is consistent, and that the use is consistent with the first
+type in ``*Ts`` which provides a relevent defintion. Type checkers MUST ONLY
+check the first relevent definition. (see rationale on ordering)
```
Comment (mark-todd) [inline docs/specification.rst:170]: relevent -> relevant (probs worth a find and replace, seen a few of these)

File: docs/specification.rst [hunk 7]
```diff
@@ -1,668 +1,481 @@
...
+short-circuiting complexity, the behavior here was not chosen for this reason,
+see rationale section for more details.
+
+Though it was not chosen for this quality, the potentially superlinear
+complexity of some other potential semantics should be kept in mind by anyone
+considering changing the semantics in the future as this may be a barrier
+to usefulness of implementations.
+
+Type-Checkers that are interested in providing tuning knobs for strictness
+MAY provide additional confiurable warnings for certain patterns with
```
Comment (mark-todd) [inline docs/specification.rst:182]: *configurable

File: docs/specification.rst [hunk 8]
```diff
@@ -1,668 +1,481 @@
...
+intersections that could cause ambiguity regarding gradual typing.. (see below)
+
+
+Composability within the type system
+------------------------------------
+
+``Intersection`` does not forbid any incompatability of type parameters
```
Comment (mark-todd) [inline docs/specification.rst:189]: *incompatibility (again, probs find and replace)

File: docs/specification.rst [hunk 9]
```diff
@@ -1,668 +1,481 @@
...
+=========
+
+Outside of the fact that we are specifying the intersections should be ordered
+in python, very little about this should be surprising. The ordering allows
+for allowing developer choice in resolving ambiguity, and comes with
+a side-effect of allowing linear complexity when evaluating
+intersections by type checkers
+
+As ordering of an intersection has the potential to be controversial,
+the below summarizes what was explored and the various tradeoffs
```
Comment (mark-todd) [inline docs/specification.rst:255]: Missing full stop

File: docs/specification.rst [hunk 10]
```diff
@@ -1,668 +1,481 @@
...
+  (See appendix below for one method of detecting potential ordering concerns)
+
+Rationale
+=========
+
+Outside of the fact that we are specifying the intersections should be ordered
+in python, very little about this should be surprising. The ordering allows
+for allowing developer choice in resolving ambiguity, and comes with
+a side-effect of allowing linear complexity when evaluating
+intersections by type checkers
```
Comment (mark-todd) [inline docs/specification.rst:252]: Missing full stop

File: docs/specification.rst [hunk 11]
```diff
@@ -1,668 +1,481 @@
...
+handled by the errors one would get if attempting to create an invalid type.
+
+While the pure and unordered form would be identical for the vast majority of
+cases we expect users of fully typed code to encounter, The combination of the
+definition of  ``Any`` doubling for both compatability and uncertainty creates
+a situation where it cannot be excluded from such an intersection, and that
+none of the possible interpretations of it are likely to be satisfying for all
+users.
+
+There are a few potential unordered interpretations of ``Any & T``
```
Comment (mark-todd) [inline docs/specification.rst:274]: Wondering if we should only include the one we're proposing in the PEP - or maybe put the others in "rejected ideas"?

File: docs/specification.rst [hunk 12]
```diff
@@ -1,668 +1,481 @@
...
+to users of a type ``Any & T`` be identical to ``Any``. This would lead to a
+significant increases in false negatives interacting with gradual typing.
+
+Banning ``Any`` from intersections to prevent this edge case would create
+significant complications for untyped imports, unbound type variables, and the
+general purpose intent of ``Any`` as a means of compatability in the framework
+of gradual typing.
+
+Other considered ways of resolving this in particular was to reword ``Any``'s
+compatability to be more flexible, and treat it as always yielding to a known
```
Comment (mark-todd) [inline docs/specification.rst:287]: *compatibility

File: docs/specification.rst [hunk 13]
```diff
@@ -1,668 +1,481 @@
...
+Neither of these provide warnings for things not provided by ``T``, the scope
+of the ordering mattering is only in the overlap.
+
+(TODO add table comparing effects of each option in each of meaningful cases, highlighting the equivalence to unordered in most cases)
+
+Backwards Compatibility
+=======================
+
+This PEP expands the allowed use of ``Self`` to better handle mixins,
+this change is not done in a backwards incompatible manner.
```
Comment (mark-todd) [inline docs/specification.rst:323]: Yeah this part I don't understand really - appreciate there is an example above, but how can classes be intersected with `Self`? Isn't `Self` just an instance of whatever it's containing class is?

File: docs/specification.rst [hunk 14]
```diff
@@ -1,668 +1,481 @@
...
+=================
+
+At the most universal level, Intersections just express that a value must
+be consistent with all of the types of the Intersection.
+This is a clear parallel to Unions, which express that a value must be
+consistent with one of the types of the Union.
+
+When teaching and documenting the edge cases that involve what happens when
+using a value that involves diamond patterns, we reccoemend drawing a parallel
+to MRO, as well as reccomending that library authors avoid creating an
```
Comment (mark-todd) [inline docs/specification.rst:348]: *recommending

File: docs/specification.rst [hunk 15]
```diff
@@ -1,668 +1,481 @@
...
+Additionally, we believe it is unlikely that all of the issues presented for
+an unordered form in the rationale section are solvable in any version of
+python that remains gradually typed, has both structural and nominal subtyping,
+and allows for resolvable diamond patterns without adversely affecting the
+needs or ergonomics for some users.
+
+Such a version of python would already likely require a python 4.0,
+allowing revisiting of both the name and operator use.
+
+Apendices
```
Comment (mark-todd) [inline docs/specification.rst:412]: Appendices

File: docs/specification.rst [hunk 16]
```diff
@@ -1,668 +1,481 @@
...
+for this which behave differently to better match the actual use cases
+their users have, one set of rules for determining this are as follows:
+
+Given any number of types, if for any identifier defined on any of the types
+there is more than 1 non-exactly equivlaent type specification for that
+identifier amoung the types, the ordering matters.
+
+Any type that is considered to be a gradual type only provides 1 definition
+for the identifiers it provides, but that 1 definition is not considered to be
+"exactly equvalent" to one provided by a non-gradual type for the purpose of
```
Comment (mark-todd) [inline docs/specification.rst:438]: *equivalent

File: docs/specification.rst [hunk 17]
```diff
@@ -1,668 +1,481 @@
...
+conflicts with ``Any`` for this check. x is a property returning Any, which is
+more strict than the behavior of ``Any`` when considering substitutability
+
+This means the inclusion of ``Any`` with a non-Any would be a reason to flag
+under this rule.
+
+However, ``Callable[..., Any]`` is also a gradual type.
+``Callable[..., Any] & SupportsAbs`` does not have multiple definitions
+for any identifier, as ``SupportsAbs`` does not provide ``__call__``
+(or any of the other things provided by ``Callable``) in an incomaptible way.
```
Comment (mark-todd) [inline docs/specification.rst:451]: *incompatible

**Review by mark-todd — COMMENTED** [review]
The spelling ones I just mark up as I see them, hope that's alright. Generally very good - seems to cover all bases of things discussed. Might be worth adding some stuff about TypedDict. Some of the phrasing I think we'll need to revisit as it's easy to get lost in the "consistent" and "expression" uses.

File: docs/specification.rst [hunk 18]
```diff
@@ -1,668 +1,481 @@
...
+
+(TODO: define or link to definitions of type expression and value expression)
+
+``A & B`` is short-hand for ``Intersection[A, B]``. ``Intersection`` is still
+needed when considering typevariable tuples.
+
+An unparameterized ``Intersection`` as a type expression must be considered an
+error by type checkers. It is possible to end up with an ``Intersection``
+that has been parameterized but is empty. This can happen in the case of type
+variable tuples. An empty intersection as a type expression is equilvant to
```
[the remark this answers is not in the document]
Comment (mikeshardmind) [reply docs/specification.rst:155]: I'll do a spelling and grammar pass in a moment to catch these at once.

File: docs/specification.rst [hunk 19]
```diff
@@ -1,668 +1,481 @@
...
+object which can be introspected consistent with the methods provided for type
+introspection in ``typing`` such as, but not limited to ``get_origin``
+
+
+Runtime type checking considerations
+------------------------------------
+
+At this point in time, ``Intersection`` as a value expression is only
+consistent with a type expression of ``object``, ``Any``, and a few internal
+types that should not be publically used.
```
[the remark this answers is not in the document]
Comment (mikeshardmind) [reply docs/specification.rst:217]: See above

File: docs/specification.rst (hunk 13 above)
Comment (mikeshardmind) [reply docs/specification.rst]: This part needs more added to explain still, haven't been adding much to the draft since noticing we have an underlying issue with unclear subtyping in the specification.

File: docs/specification.rst (hunk 11 above)
Comment (mikeshardmind) [reply docs/specification.rst:274]: Waiting to summarize each and why not for each of those until the pending discussion. It may change how compatible or not those options are.

File: docs/specification.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (mikeshardmind) [reply docs/specification.rst]: Probably needs rephrasing, will happen when back-filling in definitions used that aren't in the spec yet later.

File: docs/specification.rst [hunk 20]
```diff
@@ -1,668 +1,481 @@
...
+variable tuples. An empty intersection as a type expression is equilvant to
+``typing.Never``.
+
+An unparameterized ``Intersection`` as a value expression is not a type error,
+but may not be possible to provide a meaningful type for at this time, see
+below section on runtime typechecking support.
+
+Given an ``Intersection[*Ts]`` as a type expression, a corresponding value
+expression is considered to be consistent with it if and only if the value
+expression would be consistent with all of the type expressions, ``*Ts``.
```
[the remark this answers is not in the document]
Comment (mark-todd) [reply docs/specification.rst:165]: Isn't `TypeAlias` a valid annotation in this case, e.g.:
```py
intersect: TypeAlias = Intersection[1, 2, 3]
```

File: docs/specification.rst (hunk 5 above)
Comment (mikeshardmind) [reply docs/specification.rst:128]: I don't want to fill the pep with more examples, than specification, especially in cases where the examples would lead to "why do you even need this?", the question posed in that stack overflow question would benefit from protocols over multiple inheritance + still needing to implement the methods anyway. That example would still want to be able to specify multiple applicable protocols, but it isn't a good example of mixin behavior.

File: docs/specification.rst (hunk 4 above)
Comment (mikeshardmind) [reply docs/specification.rst]: I added details about kotlin's being non-denotable. let me know if that's enough additional clarity for what you were thinking in that section

File: docs/specification.rst (hunk 5 above)
Comment (mark-todd) [reply docs/specification.rst:128]: Yeah I do see that, I'm not saying that it should be crammed full of examples, and I think having a practical one is very handy. Maybe what I'm describing belongs more in the "how to teach" this section, but I guess I'm thinking of the reader who may not be at all familiar with intersections. I think the following would be quite a simple motivating example:

```py

from abc import ABC

# We have two interface classes
class AbstractA(ABC):
    def foo(self) -> int:
        ...

class AbstractB(ABC):
    def bar(self) -> str:
        ...

# We have a class that inherits from both
# but not directly
class A(AbstractA):
    def foo(self):
        return 1

class B(AbstractB):
    def bar(self):
        return "test"

class C(A,B):
    pass

# Without intersection there is no way to describe a function that takes
# something that has the methods of C, but may not inherit
# directly from C
def test(x: AbstractA & AbstractB):
    return x.foo(), x.bar()
```

Comment (mark-todd) [reply docs/specification.rst:128]: This might seem like a bit of a toy example, but I have seen projects that use these kind of structures - typically the defined methods `foo` and `bar` here can be very complex, so breaking them up across multiple files might be a good idea, but this can lead to issues like above.

File: docs/specification.rst (hunk 20 above)
[the remark this answers is not in the document]
Comment (mikeshardmind) [reply docs/specification.rst:165]: Yes, but it isn't consistent with it as solely a value. This is a special case where the value holds a type expression, and `TypeAlias` has special behavior. In new python versions and code with relatively new minimum python specificed, you'd have this equivalently instead, where it's more obvious that it's a type expression. 

```py
type intersect = Intersection[1, 2,3]
```

There are a few other places where a value must be a type expression, such as the second parameter to `assert_type` and `cast`, and there's ongoing discussion here: https://discuss.python.org/t/basic-terminology-for-types-and-type-forms/46741 for documenting all of this. and there's a need for a TypeForm/SpecialForm that's got some people wanting it discussed some on mypy's tracker here: https://github.com/python/mypy/issues/9773

**Comment by mark-todd** [comment]
> @mark-todd discussion restarted at https://discuss.python.org/t/take-2-rules-for-subclassing-any/47981 per Eric's comment on the immediate back and forth being harder to follow

@mikeshardmind It seems from the discussion here that there's quite large support for the "LSP Compatible Any" option - I would suggest maybe we move forward assuming this is the preferred outcome? Then if it later turns out there's resistance to this it could be adjusted

**Comment by mark-todd** [comment]
> Given my available time right now, I'll wait until we have a clear direction. The sections that need the most detail added to them next are the ones that would have the most changed by a decision either way.

@mikeshardmind Yeah fair enough, I guess what I meant was I'm not sure a clearer decision will materialise (although I could be wrong) - but equally if things are busy for you at the moment no harm in leaving it for a bit and seeing what happens

**Comment by mikeshardmind** [comment]
I think the main thing we're waiting on for "LSP Compatible Any" to be a viable path forward is an expression of the idea that doesn't break language server uses of typing and that is consistent. I can likely still make some progress on this draft this weekend since behaviorally both options are the same, but it's hard to flesh out a lot of the how to teach it and rationale without having picked the underlying reason we picked that determines that behavior.

I can also put more effort into assisting with coming up with such a phrasing, but won't be able to for the next few days.

**Comment by mark-todd** [comment]
> I think the main thing we're waiting on for "LSP Compatible Any" to be a viable path forward is an expression of the idea that doesn't break language server uses of typing and that is consistent. I can likely still make some progress on this draft this weekend since behaviorally both options are the same, but it's hard to flesh out a lot of the how to teach it and rationale without having picked the underlying reason we picked that determines that behavior.

> I can also put more effort into assisting with coming up with such a phrasing, but won't be able to for the next few days.

Sure makes sense - I suppose I would argue that "how to teach it" could be probably be done by example, if behaviourally they are each the same (or at least that's how I'd probably teach it), but rationale would require the reason to be known.

On a side note I'm thinking of updating the "intersect simulator" to follow the new proposal, my thought is we might be able to answer a lot of potential questions by getting people to try it out in practice.

**Comment by mark-todd** [comment]
I've now implemented the intersect simulator for the new proposal https://github.com/CarliJoy/intersection_examples/pull/34

Only part I haven't figured out yet is how to deal with Callables, but I now think this is probably because I'm not sure it's been fully fleshed out yet. Conceptually it makes sense - if you request the `__call__` method, the first one in the intersection takes priority. Issue is, if you try and access `__call__` from a Callable, it doesn't actually use it like a class or protocol - e.g.:

```py
def test(i: int) -> str:
    return str(i)

x: Callable[[int], str] = test
y = x.__call__(1) # raises an error in pyright
```

Just bringing it up as this PR is now closed: https://github.com/CarliJoy/intersection_examples/issues/16

It seems to me that `typing.Callable` and `__call__` methods are somewhat incompatible - am I missing something here?

e.g. what is the type of `x` and `y` in the following:

```py
class A(typing.Protocol):
    def __call__(self, x: int) -> str:
         ...

a: Intersection[Callable[[str], int], A]
b: Intersection[A, Callable[[str], int]]

x = a.__call__
y = b.__call__
```

**Comment by CarliJoy** [comment]
```
class A(typing.Protocol):
    def __call__(self, x: int) -> str:
```

Please note that this object is **not** intented to be called with `.__call__` but rather `a()` as is.
I don't think we need to consider this here.

therefore `x = a.__call__` is not set in both cases IMHO.

**Comment by mark-todd** [comment]
> Please note that this object is not intented to be called with .__call__ but rather a() as is.
> I don't think we need to consider this here.

> therefore x = a.__call__ is not set in both cases IMHO.

@CarliJoy I realise now that my example was perhaps unclear - firstly I believe that these two are equivalent:

```py
a = A()
a.__call__(1) # Usage 1
a(1) # Usage 2
```
Although usage 2 is the intended usage, I believe the two are treated as the same.

More importantly though, I was attempting to raise a point that affects the intended usage as well. Consider instead:
```py
class A(typing.Protocol):
    def __call__(self, x: int) -> str:
         ...

a: Intersection[Callable[[str], int], A]
b: Intersection[A, Callable[[str], int]]

a(...) # Label 1
b(...) # Label 2
```
The question really is, what does the type checker predict at Label 1 vs Label 2? Does the intersection with Callable overwrite the `__call__` method of A?

**Comment by mikeshardmind** [comment]
Not really a concern of the pep, type checkers need to understand callables and callable protocols already. Your protocol is incompatible with the callable provided, as callables are positional only so there won't ever be anything valid to pass to that intersection.

**Comment by mark-todd** [comment]
Ok so I'm not sure how much of this is just about the way VS code and pyright interpret things, but here's the example:

```py
import typing

class A(typing.Protocol):
    def __call__(self, x: int, /) -> str:
         ...

class B:
    def __call__(self, x: int, /) -> str:
        return str(x)

def func(y: int) -> str:
    return str(y)

def usage_func(a: A, b: typing.Callable[[int], str]):
    a(1)
    b(1)
    a.__call__(1)
    b.__call__(2)

usage_func(B(), func)
```
Critically, when you hover over a(1), it displays like this, with the name `x` specified:

image

But then b(1) displays like this, with simply the type:

image

And also the `__call__` method is unknown on b. My point is that unfortunately `A` and `typing.Callable[[int], str]` are not identical - I think we need to resolve if one takes priority over the other, and if the fact that `__call__` is not present on the `Callable` type matters. I know it's not the intended use case, but it is still a method that's on one and not the other.

Personally I think we resolve it this way:

* We treat `__call__` as being being present on `A` but not `typing.Callable[[int], str]`, so even if later in intersection ordering A takes priority
* We treat the actual call (with brackets) as being present on both
* The "style" of the displayed method depends on which takes priority in the intersection (a.k.a. if A comes first the label `x` appears, and if the `Callable` comes first it appears with no label)

**Comment by mikeshardmind** [comment]
This seems like a pile of typechecker-specific behavior, and something we specifically *shouldn't* specify.

`Callable[[T], R]` has the same structure as 

```py
class Callable(Protocol):
    def __call__(self, arg: T, /) -> R:
        ...
```

type checker not synthesizing a `.__call__` for Callable isn't really our problem and should be brought up with type checkers. Same for the parameter names being displayed as part of the type for positional-only parameters.

**Comment by mikeshardmind** [comment]
Data Model shows off relative equivalence: https://docs.python.org/3/reference/datamodel.html#object.__call__

[Typing spec says ](https://typing.readthedocs.io/en/latest/spec/callables.html#callback-protocols)

> Callback protocols and Callable[...] types can be used interchangeably.

**Comment by mark-todd** [comment]
> This seems like a pile of typechecker-specific behavior, and something we specifically shouldn't specify.

Maybe you're right - the line between type checker and PEP responsible is often difficult to discern.

> Same for the parameter names being displayed as part of the type for positional-only parameters.

This part I think I agree with, based on the link to the docs you posted. What gets displayed seems like something the type checkers can resolve

> type checker not synthesizing a .__call__ for Callable isn't really our problem and should be brought up with type checkers.

I'm starting to this one may actually be a bug in the PEP specification that's already been accepted. What the type checker is displaying here is correct - there is no `__call__` method on the function (it'll fail at runtime). And yet classes accept `__call__` usage, via protocol or otherwise. I don't see how they can be fully interchangable unless:
* Usage of `__call__` is banned
* `__call__` maps to calling in functions as well somehow, but that would be weird because it would kind of map to itself.

**Comment by mark-todd** [comment]
This means the styling issue we can ignore - that's up to type checkers - but pending some change to the way `__call__` works I think we have to keep these two. I suspect these already emerge from the current specification anyway, I'm just spelling it out really:

* We treat `__call__` as being being present on `A` but not `typing.Callable[[int], str]`, so even if later in intersection ordering A takes priority
* We treat the actual call (with brackets) as being present on both

**Comment by mikeshardmind** [comment]
> I'm starting to this one may actually be a bug in the PEP specification that's already been accepted. What the type checker is displaying here is correct - there is no __call__ method on the function (it'll fail at runtime). 

No, it exists on function objects at runtime too. Type checkers rejecting direct use of `.__call__` on function objects doesn't match runtime.

```py
>>> def func(y: int) -> str:
...     return str(y)
...
>>> func.__call__(1)
'1'
>>>

**Comment by mark-todd** [comment]
> No, it exists on function objects at runtime too. Type checkers rejecting direct use of .__call__ on function objects doesn't match runtime.

Cool! Yeah we're all good then that's great - apologies I thought I'd tested this but apparently not.

> I suspect that type checkers might have to actually fix this to implement intersections on callable types, but I don't think we need to specify anything additional for this case.

Yeah, I wonder if it's worth flagging somewhere, not sure if that belongs in the PEP but maybe an issue after acceptance. Did have one thought - maybe we should a section for "Type Checker Pre-requisites" that highlights things that might need to be fixed in type checkers in order for intersection to work? I think some other conversations about the way Any/LSP works may already fall into this category.

**Comment by mikeshardmind** [comment]
> Yeah, I wonder if it's worth flagging somewhere, not sure if that belongs in the PEP but maybe an issue after acceptance. Did have one thought - maybe we should a section for "Type Checker Pre-requisites" that highlights things that might need to be fixed in type checkers in order for intersection to work? I think some other conversations about the way Any/LSP works may already fall into this category.

Based on the current conformance suite and the prerequisite Any subtyping behavior, I agree, but I'm still thinking through handling of that in terms of how to say it in the PEP, I'd like to take some time in one of the next few days to get this to a state we can present to a core dev to see about finding a sponsor, an actual implementation does not need to be ready at the draft stage.

I'm sort of assuming based on the state of the discussion on Any that there's negative appetite on changing current type checker behavior on *Existing* subtyping with Any, and there's good reason for it, so I think we can expand on specifying the needed behavior as part of this PEP rather than separately since we aren't requiring a prerequisite change, only that going forward we have a consistent definition.

**Comment by DiscordLiz** [comment]
> so I think we can expand on specifying the needed behavior as part of this PEP rather than separately since we aren't requiring a prerequisite change, only that going forward we have a consistent definition.

I'd appreciate if you have a way to work in the fact that there are multiple interpretations which each reach that same conclusion, but I'd understand if that doesn't fit into the PEP's scope. I won't be able to actively participate as much for a few weeks, but you can reach me if you want my opinion on any of this.
